roaringbitmap/src/main/java/org/roaringbitmap/RoaringBitmapWriter.java

```diff
@@ -0,0 +1,283 @@
+package org.roaringbitmap;
+
+import org.roaringbitmap.buffer.*;
+
+import java.util.function.Supplier;
+
+public interface RoaringBitmapWriter<T extends BitmapDataProvider> extends Supplier<T> {
+
+  static Wizard<Container, RoaringBitmap> writer() {
+    return new RoaringBitmapWizard();
+  }
+
+  static Wizard<MappeableContainer, MutableRoaringBitmap> bufferWriter() {
+    return new BufferWizard();
+  }
+
+  abstract class Wizard<C extends WordStorage<C>,
+          T extends BitmapDataProvider & AppendableStorage<C>>
+          implements Supplier<RoaringBitmapWriter<T>> {
+
+    protected int initialCapacity = RoaringArray.INITIAL_CAPACITY;
+    protected boolean constantMemory;
+    protected boolean partiallySortValues = false;
+    protected Supplier<C> containerSupplier;
+
+    Wizard() {
+      containerSupplier = arraySupplier();
+    }
+
+    /**
+     * Choose this option if it is known that most containers will be sparse.
+     * @return this
+     */
+    public Wizard<C, T> optimiseForArrays() {
+      containerSupplier = arraySupplier();
+      return this;
+    }
+
+    /**
+     * Choose this option if the bitmap is expected to be RLE compressible.
+     * Will buffer additions into a RunContainer.
+     * @return this
+     */
+    public Wizard<C, T> optimiseForRuns() {
+      containerSupplier = runSupplier();
+      return this;
+    }
+
+    /**
+     * Choose this option if the bitmap is expected to be mostly dense but unlikely to be RLE
+     * compressible
+     * @return this
+     */
+    public Wizard<C, T> optimiseForBitmaps() {
+      containerSupplier = bitmapSupplier();
+      return this;
+    }
+
+    /**
+     *
+     * @param count how many values are expected to fall within any 65536 bit range.
+     * @return this
+     */
+    public Wizard<C, T> expectedValuesPerContainer(int count) {
+      sanityCheck(count);
+      if (count < ArrayContainer.DEFAULT_MAX_SIZE) {
+        return optimiseForArrays();
```

Comment:
wouldn't it be great if `count` affect the array allocated in `ArrayContainer` supplier?

Reply:
ArrayContainer::new allocates array of length 4. I'm thinking about allocating` new int[count]`.

Reply:
Yes that's a quick win.